### PR TITLE
feat: generate immutable types

### PR DESCRIPTION
**Pull request by dnalborczyk**
fixes: https://github.com/drwpow/openapi-typescript/issues/501

### CODE REVIEW

File: bin/cli.js [hunk 1]
```diff
@@ -15,6 +15,7 @@ Options
   --help                display this
   --output, -o          Specify output file (default: stdout)
   --auth                (optional) Provide an authentication token for private URL
+  --immutableTypes      (optional) Generates immutable types (readonly properties and readonly array)
```
Comment (drwpow) [inline bin/cli.js:18]: nit: `meow` allows camelCase or kebab-case flags; WDYT about just using kebab-case here? Both should be supported behind-the-scenes

File: src/transform/headers.ts [hunk 2]
```diff
@@ -1,15 +1,23 @@
...
+  { immutableTypes }: { immutableTypes: boolean }
+): string {
   let output = "";
 
   Object.entries(headerMap).forEach(([k, v]) => {
     if (!v.schema) return;
 
     if (v.description) output += comment(v.description);
+
+    const readonly = immutableTypes ? "readonly " : "";
```
Comment (drwpow) [inline src/transform/headers.ts:15]: 👍🏻

**Comment by drwpow** [comment]
Awesome! Love the direction of this PR; I would have done it the same way.

My only suggestion would be to add at least 1 test for this. But the nice part of adding a new feature is that we’re not breaking behavior for anybody. So even if we missed a type, we could add it after-the-fact.

File: src/transform/headers.ts (hunk 2 above)
Comment (drwpow) [reply src/transform/headers.ts:15]: I personally am also a fan of `readonly string[]` over `ReadonlyArray<string>`. AFAIK this looks easier to implement, and has no downsides. I’m in favor of this approach.

File: bin/cli.js (hunk 1 above)
Comment (dnalborczyk) [reply bin/cli.js:18]: oh, yeah, definitely. I didn't even notice `prettier-config` and `raw-schema` being `kebab-case`.

**Comment by dnalborczyk** [comment]
> Awesome! Love the direction of this PR; I would have done it the same way.

Thank you! 😊 

> My only suggestion would be to add at least 1 test for this.

yeah, I'll add some tests. just wanted to see the current tests passing.

> But the nice part of adding a new feature is that we’re not breaking behavior for anybody. So even if we missed a type, we could add it after-the-fact.

I agree.

File: src/transform/headers.ts (hunk 2 above)
Comment (dnalborczyk) [reply src/transform/headers.ts:16]: the "downside" - if you want to call it such - would be a minimum requirement of TypeScript v3.4, which I think should be fine considering v4.2 is currently latest.

Comment (drwpow) [reply src/transform/headers.ts:16]: TS 4 was such an easy update I don’t know of any reason why people would still be on an old version. But given that this is opt-in too, this is more than fine.

**Comment by dnalborczyk** [comment]
@drwpow the PR is pretty much good to go. I just want to merge some duplicate interfaces I created for passing around options: { immutableTypes }. gonna try to get to it tomorrow.

**Comment by drwpow** [comment]
Awesome. I’ll take a look tomorrow. It’s looking good to me already.

File: src/transform/index.ts [hunk 3]
```diff
@@ -7,11 +7,14 @@ import { transformResponsesObj, transformRequestBodies } from "./responses";
...
 
 interface TransformOptions {
+  immutableTypes: boolean;
   rawSchema?: boolean;
   version: number;
 }
 
-export function transformAll(schema: any, { version, rawSchema }: TransformOptions): string {
+export function transformAll(schema: any, { immutableTypes, rawSchema, version }: TransformOptions): string {
+  const readonly = immutableTypes ? "readonly " : "";
```
Comment (drwpow) [inline src/transform/index.ts:16]: Not a blocker for this PR, but I’d suggest moving this into `src/utils.ts` as `tsReadonly()` util. Similar to the existing `tsUnionOf()` util and others.

It’s a minor thing, but it makes modifications quicker, and helps prevent typos.

File: tests/operation.test.ts [hunk 4]
```diff
@@ -72,5 +99,22 @@ describe("requestBodies", () => {
...
+          /** Pet request body */
+          Pet: {
+            readonly content: {
+              readonly "application/json": {
+                readonly test?: string;
+              };
+            };
+          };
+        };`)
+    );
```
Comment (drwpow) [inline tests/operation.test.ts:118]: 💯

File: tests/v3/index.test.ts [hunk 5]
```diff
@@ -1,18 +1,33 @@
...
         cwd: __dirname,
       });
-      const expected = fs.readFileSync(path.join(__dirname, "expected", output), "utf8");
-      const generated = fs.readFileSync(path.join(__dirname, "generated", output), "utf8");
+      const expected = readFileSync(join(__dirname, "expected", output), "utf8");
+      const generated = readFileSync(join(__dirname, "generated", output), "utf8");
+      expect(generated).toBe(expected);
+    });
+
+    it(`reads ${schema} spec (v3) from file (immutable types)`, () => {
```
Comment (drwpow) [inline tests/v3/index.test.ts:20]: Love all the immutable tests!

**Review by drwpow — APPROVED** [review]
This PR is looking good! I have no changes requested. Ping me when you feel you’re happy with it and I can merge.

I can also either release an RC if you’d like to do more testing, or we can just cut a minor release. Your call. Again, since this is a new flag, I’m not as concerned with just releasing because we won’t be breaking anyone.

File: src/transform/index.ts (hunk 3 above)
Comment (dnalborczyk) [reply src/transform/index.ts:16]: no problem, I can do that quick as well. I had it set up originally as a function, but changed it back.

**Comment by dnalborczyk** [comment]
> This PR is looking good! I have no changes requested. Ping me when you feel you’re happy with it and I can merge.
> 
> I can also either release an RC if you’d like to do more testing, or we can just cut a minor release. Your call. Again, since this is a new flag, I’m not as concerned with just releasing because we won’t be breaking anyone.

should be good to go! 👍 **minor** release should be fine 🤞

**Comment by dnalborczyk** [comment]
@drwpow just want to point out a couple things quick:

**Comment by drwpow** [comment]
Thanks so much! Agree with all of these changes. Thanks again for contributing

**Comment by drwpow** [comment]
@all-contributors please add @dnalborczyk for code, test

**Comment by dnalborczyk** [comment]
> Thanks so much! Agree with all of these changes. Thanks again for contributing

@drwpow  thank you for this great project and thank you for accepting the PR!

**Comment by laurence-myers** [comment]
@drwpow @dnalborczyk Thanks for implementing & reviewing this feature! I really appreciate your time & efforts. 🎉 

(I just wanted to join the "thanks" party! 😁)

**Pull request merged**